cdap-api/src/main/java/co/cask/cdap/api/workflow/WorkflowContext.java

```diff
@@ -51,4 +52,10 @@
    * @return a {@link WorkflowToken}
    */
   WorkflowToken getToken();
+
+  /**
+   * @return the {@link WorkflowActionContext} for custom actions
+   */
+  @Nullable
+  WorkflowActionContext getWorkflowActionContext();
```

Comment:
I did it like this, because not all Workflow's will have a custom action.  In those cases, getWorkflowActionContext will just return null.  Otherwise, the WorkflowActionContext will have the datasets from WorkflowActionSpecification.getDatasets() that AbstractContext can instantiate.  If a Workflow does not have a custom action, we shouldn't need to pass in all the DatasetFramework, TransactionSystemClient, DiscoveryServiceClient etc.. right? (constructor of BasicWorkflowActionContext)